crates/proto/src/rr/rdata/tsig.rs

```diff
@@ -0,0 +1,770 @@
+    let other =
+        decoder.read_vec(other_len as usize)?.unverified(/*valid as any vec ot the right size*/);
+
+    Ok(TSIG {
+        algorithm,
+        time,
+        fudge,
+        mac,
+        oid,
+        error,
+        other,
+    })
+}
+
+/// Write the RData from the given Encoder
+///
+/// ```text
+///                       1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3
+///   0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  /                         Algorithm Name                        /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |                                                               |
+///  |          Time Signed          +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |                               |            Fudge              |
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |          MAC Size             |                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+             MAC               /
+///  /                                                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |          Original ID          |            Error              |
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |          Other Len            |                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+           Other Data          /
+///  /                                                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// ```
+pub fn emit(encoder: &mut BinEncoder<'_>, tsig: &TSIG) -> ProtoResult<()> {
+    tsig.algorithm.emit(encoder)?;
+    encoder.emit_u16((tsig.time >> 32) as u16)?;
+    encoder.emit_u32(tsig.time as u32)?;
+    encoder.emit_u16(tsig.fudge)?;
+    encoder.emit_u16(tsig.mac.len() as u16)?;
+    encoder.emit_vec(&tsig.mac)?;
+    encoder.emit_u16(tsig.oid)?;
+    encoder.emit_u16(tsig.error)?;
+    encoder.emit_u16(tsig.other.len() as u16)?;
+    encoder.emit_vec(&tsig.other)?;
+    Ok(())
+}
+
+// Does not appear to have a normalized text representation
+impl fmt::Display for TSIG {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        write!(
+            f,
+            "{algorithm} {time} {fudge} {mac} {oid} {error} {other}",
+            algorithm = self.algorithm,
+            time = self.time,
+            fudge = self.fudge,
+            mac = sshfp::HEX.encode(&self.mac),
+            oid = self.oid,
+            error = self.error,
+            other = sshfp::HEX.encode(&self.other),
+        )
+    }
+}
+
+impl Algorithm {
+    /// Return DNS name for the algorithm
+    pub fn to_name(&self) -> Name {
+        use Algorithm::*;
+        match self {
+            HmacMd5 => Name::from_ascii("HMAC-MD5.SIG-ALG.REG.INT"),
```

Comment:
can you verify case here? I just notice all the others are lowercase... Also, given the DNS label conversion here, can you confirm if this should be fqdn, i.e. ending with a `.`. For example: `HMAC-MD5.SIG-ALG.REG.INT.` and `gss-tsig.`

Reply:
The case is correct (see table in https://datatracker.ietf.org/doc/html/rfc8945#section-6).
I believe they should not be fqdn, otherwise I would expect the rfc to add this final dot.
The only place talking about this looking like a fqdn is here https://datatracker.ietf.org/doc/html/rfc2845#section-7, and it's not helping me either way.
> The algorithm name HMAC-MD5.SIG-ALG.REG.INT
   looks like a FQDN for historical reasons; future algorithm names are
   expected to be simple (i.e., single-component) names.

Reply:
perfect, thanks.